src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java

```diff
@@ -300,17 +308,45 @@ public void publish(ClusterState clusterState, final Discovery.AckListener ackLi
         return members.toArray(new LocalDiscovery[members.size()]);
     }
 
-    private void publish(LocalDiscovery[] members, ClusterState clusterState, final BlockingClusterStatePublishResponseHandler publishResponseHandler) {
+    private void publish(LocalDiscovery[] members, ClusterChangedEvent clusterChangedEvent, final BlockingClusterStatePublishResponseHandler publishResponseHandler) {
 
         try {
             // we do the marshaling intentionally, to check it works well...
-            final byte[] clusterStateBytes = Builder.toBytes(clusterState);
+            byte[] clusterStateBytes = null;
+            byte[] clusterStateDiffBytes = null;
 
+            ClusterState clusterState = clusterChangedEvent.state();
             for (final LocalDiscovery discovery : members) {
                 if (discovery.master) {
                     continue;
                 }
-                final ClusterState nodeSpecificClusterState = ClusterState.Builder.fromBytes(clusterStateBytes, discovery.localNode);
+                ClusterState newNodeSpecificClusterState = null;
+                discovery.lock.lock();
```

Comment:
can we use ReleasableLock and try with resources? (or a plain old synchronize :) )